### ISSUE TITLE
String literals are not escaped properly when no source code is provided for formatting

### ISSUE THREAD

**Issue by brianberns** [fsprojects/fantomas#1546]
## Description
When `None` is passed as the `source` to `CodeFormatter.FormatASTAsync`, the resulting formatted F# code contains incorrectly escaped string literals.

## Reproduction
Unit test:
```
let myFormatSourceString isFsiFile (s: string) config =

    let fileName =
        if isFsiFile then
            "/src.fsi"
        else
            "/src.fsx"

    let parsingOptions =
        { FSharp.Compiler.SourceCodeServices.FSharpParsingOptions.Default with
            SourceFiles = [| fileName |] }

    async {
        let! pairs =
            Fantomas.CodeFormatter.ParseAsync(
                fileName,
                Fantomas.SourceOrigin.SourceString s,
                parsingOptions,
                sharedChecker.Value)
        let ast, _ = Array.exactlyOne pairs
        let! formatted =
            Fantomas.CodeFormatter.FormatASTAsync(
                ast,
                fileName,
                List.empty,
                None,    // *** was: Some (SourceOrigin.SourceString s)
                config)

        return formatted.Replace("\r\n", "\n")
    }

    |> Async.RunSynchronously

[<Test>]
let ``string literals`` () =
    myFormatSourceString
        false
        """
let str2 = "c:\\test"
    """
        ({ config with
               MaxValueBindingWidth = 60 })
    |> prepend newline
    |> should
        equal
        """
let str2 = "c:\\test"
"""
```

Result:
```
  Message: 
      Expected string length 23 but was 22. Strings differ at index 16.
      Expected: "\nlet str2 = "c:\\\\test"\n"
      But was:  "\nlet str2 = "c:\\test"\n"
      -----------------------------^
```

## Cause

I tracked the problem down to [the logic in CodePrinter.fs](https://github.com/fsprojects/fantomas/blob/df92d9f8831a3fac0139643084c7e728ea9087a8/src/Fantomas/CodePrinter.fs#L5367) that handles `SynConst.String`. When no trivia are present, the code successfully escapes double-quote characters, but does not escape backslash characters.

## Suggested fix

Add the following line, or the equivalent:

```
let escaped = Regex.Replace(escaped, @"\\", @"\\")
```

See [this StackOverflow question](https://stackoverflow.com/questions/6105476/replace-single-backslash-with-double-backslash) for details.

**Comment by nojaf** [comment]
Just to be sure, what is your use-case for using `CodeFormatter.FormatASTAsync`?
The next version of FCS will allow more information to be stored in AST, so this would be easy to once that lands.

See https://github.com/dotnet/fsharp/blob/dbb8d47059ac1459c00b1ea84e6e2dbcde4b16a9/src/fsharp/SyntaxTree.fsi#L69-L74

Duplicate of https://github.com/fsprojects/fantomas/issues/560

**Comment by brianberns** [comment]
Thanks for the quick reply. I'm working on a C#-to-F# translator, so I'm building an F# AST from a C# AST, and then sending that AST through Fantomas. I'm not using the F# compiler at all, so there's no F# source code to send into `CodeFormatter`.

**Comment by nojaf** [comment]
Thanks for the details on your use-case.
In this case, you also want to set[ StrictMode = true](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#fsharp_strict_mode).
[Sample](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAbRAXMBnLATgExgC8YAOjChbRVovlZADThS7ZYCWAdgOa5UYANrsIoCFKgASLAE8ADsjSQ%2BWVuKmQZANwCG6AK4rRW6QAY206ZACSvJLywBlbgC9k1mxAAs5sABdLQBfbwkAnQVlYTVnTR8dA2NTMR8IAEZwm0gAWX0ADwAZPkQixAEsAAsE9IyiCwDg6TCtSRy5JVNIACNYWHRa2z1DE2E09KJs2xdEAFtuOHRYXgBBLABRRwB5ADMS3i8AiF3DTibQ7Pbh6O6%2BgaHtEZTx47AAZmntF0V9aEQAEKIXawAiIAAK%2BgI%2Bjm2EQBEe0kIJguLSukU6MVU90GXySo1Sb18XwgkB%2Bf0BwNBZVgAHd4dB9JwHLpYIyeCtEVJkYhUVJWtJrk9brEcVz8S80BMfABWElQcn-IEgsEAVUUygIjOZvFZ7O4nPlp3Q5x8zX56MSmLu-VxGOSYylbwAbPKyb8lVSwQBhdBM3Delb4AhGaBYUFck5nXlmy5tDEi7G28XPR1mdJgADsbsVlJViFy8x68MjYGNpps5ogAqkQtJ1tFybxqcJGYAHDmPXnqYHlrxS%2BWY5W44KE11Gw9mw7W%2BkAJydimrXYMAiBuZzfSlnl86uWjqJqBiqcE14ZjJWN7uinK6mzBZLQ1vQc7sA1iJWg%2B9Jv2k9Os9ZS9cyXFc70WAZHwzbdYzReMP3HJNJx-SV0zqKZLwcJwsG2XgABUCHkAB1bhqgHaMXzfMA61kT8jyQtNpRsDJPkArtVgIWAjEcAARRB0G4BYVy3EMh2kKtXz3G54KgdQU2nU86mJS98gKOxdhwqoKg2E1EBcKpQSwIj4BI%2BVfEaaCLVg-cpLiDRj2QhjpAyOUlMKBxdm4Aptk1fRwwIDYCkUMFcFwA1%2B3lGUzOHGDRzgrFpPiOz6LeDJXRcgoACVEDgAh4EM4yiUi0SR1rMc4ps2TfxQnwMmzNLMuy%2BAADkjDmYsCD2OwGDmIR5QyciJOFay5iMdAeD4w4ADFQQ3LAVxw8dEpnaqO0verQXgXIRrG0opoIGbBLdaAqihP4VwAfVpbgjJqfrLMksqZMW%2BTqvnNK2OheRtgIEp8DymoTMKqQxIoqiomsx66KWmwGjdZT3v0T7vu4fBmta%2BEOq6nrktumKrLK4bRu4cbEF2-b4XmmInr-SYAIzSB4cRn6sE2wnidJnyDsvI6TrDeELqukicZK2Lughq05OpnwiDQunlIANQJAE%2BHgPh%2BD%2B0tTKF988dFhLIee6HmNlwoJs4sNQqVxxVfVgGtco0rdds-XJehxTjYKLjYCwABxbB-MCxgQpWG23giu3QYbBC7XFyqHKkIhnNlraidKAFlmgABrAFoUz7BcGwlwYUQXsWrCp8yPM3c7sGsraJj%2By3iIVK6caxBaWJoEsHpCoKcQHj3N4YjQtWRxCzRggsYzZ9K-E6v6xo7966SjMiFqumAGlEEQRRVPUioHEL2EDiOKeK6iizcfux2KobleVvdrSFlwKoQ4zTWZ5Bh2J2jjoJaq6HXp0zcAITAbF%2BAtQqAZRAPRcgIBPukae58q6XxrjaRCS8obSHeBeOmqw%2BL8F4KbXg5sVhuAIT5IwYIcKwAwpAnyoVSLaXDl-KON9l7pHeLTdI9NRrwl4PQ3QNIBBjzav3PgQ8gyMIrEVaKwsdbfzYZgqQ7wZbcJZjwY%2BRQYQ9HgPoX0sAQoCFbu3UoUiRJA2KtrK%2BCiqb-ywUbbhXFkb6B6JgR%2ByMqguHkM4QoZjmEi1iBUeAexj4uAUJgRRBssFu24VsYJ%2BxTGHQIOgXYkB-HyNYbYuOEB3iJ24ZvbetDnAOGzvoYh-1y5MI-gNee1k66-1jm8d4zduFhIIIsZmcChIog-uwQI3g1C4AmiFYQz4QhAA)

The unit test you want in this case is
```fsharp
[<Test>]
let ``escape slash in string using strict mode, 1546`` () =
    formatSourceString
        false
        """
let str2 = "c:\\test"
"""
        { config with StrictMode = true }
    |> prepend newline
    |> should
        equal
        """
let str2 = "c:\\test"
"""
```

in [SynConstTests](https://github.com/fsprojects/fantomas/blob/df92d9f8831a3fac0139643084c7e728ea9087a8/src/Fantomas.Tests/SynConstTests.fs).

I would accept a PR with your suggested fix.
Are you interested?
If so, please read our [Contribution Guidelines](https://github.com/fsprojects/fantomas/blob/master/CONTRIBUTING.md#pull-request-ground-rules).

**Comment by brianberns** [comment]
Sure, I'll submit a PR. Thanks for the suggestion to use StrictMode.

**Comment by brianberns** [comment]
Unfortunately, my proposed fix breaks unit test `quotes should be escaped in strict mode`, because it incorrectly attempts to escape the backslash in `\"`. I think someone with better regex skills than me will have to take a look at this problem.

```
  Message: 
    System.Exception : The formatted result is not valid F# code or contains warnings
    let formatter = sprintf "%i,\\"%s\\""
```

**Comment by nojaf** [comment]
Hmm, would it be easier perhaps to transforms the string in your tool so that is it double escaped there?

**Comment by brianberns** [comment]
Yes, I can work around the problem in my code for now.

**Comment by nojaf** [comment]
I'm going to close this, feel free to open a new issue if you think there is any action left here.
